python/ql/lib/semmle/python/frameworks/Cryptography.qll

```diff
@@ -182,7 +193,12 @@ private module CryptographyModel {
               .getACall() and
         algorithmClassRef(algorithmName).getReturn().getAUse() in [
             call.getArg(0), call.getArgByName("algorithm")
-          ]
+          ] and
+        exists(DataFlow::Node modeArg | modeArg in [call.getArg(1), call.getArgByName("mode")] |
+          modeArg = modeClassRef(modeName).getReturn().getAUse()
+          or
+          modeArg.asExpr() instanceof None and modeName = "<none>"
```

Comment:
In general I don't find QL very good at tracking multiple properties of instances, so I understand why you wrote the code as you did. However, if we for some reason don't manage to match either part of the disjunct for the `modeName`, we also loose everything about the algorithmName :disappointed: 

I see that some of the [official examples](https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/#cryptography.hazmat.primitives.ciphers.algorithms.ChaCha20) use `mode=None`, so good catch on handling that as well :+1:

I would feel better with a defensive version like one below:

```suggestion
          if modeArg = modeClassRef(modeName).getReturn().getAUse()
          then any()
          else modeName = "<None or unknown>"
```

it does feel slightly hacky to have this `<None or unknown>` instead of just not having a result... I guess this could be fixed by having two sets cipher instance tracking (one for algorithm name, one for mode) -- but then you might end up with cross-talk between unrelated algorithm and mode pairs :thinking:

_note: I haven't actually tested this code, but I hope it "should just work" :crossed_fingers:_

Reply:
Thanks, that's a good point. I think I prefer this fallback case over trying to track the two things separately for the ease of keeping the algorithm and mode bundled together. The slight hackiness seems acceptable as the `modeName` itself shouldn't be exposed to users of the `CryptographyGenericCipherOperation` class as this is exposed via the `Cryptography::BlockMode getBlockMode()` predicate, which is restricted to only known block modes.